File: lib/features/activity_backcountry_ski/models/backcountry_ski_conditions_report.dart
import 'package:turbo/features/activity_fishing/api.dart' show WeatherSlice;

/// Typed conditions report returned by
/// `/api/activities/backcountry-ski/{id}/conditions`. Mirrors the
/// server's [BackcountrySkiConditionsReport]. Reuses the
/// [WeatherSlice] exported from the fishing kind — same data shape,
/// same client-side parser. Each kind owns its scoring + rationale.
class BackcountrySkiConditionsReport {
  final String activityId;
  final DateTime validAt;
  final DateTime fetchedAt;
  final WeatherSlice weather;

  /// Varsom avalanche danger level 1–5 if available. Null until the
  /// VarsomAvalancheProvider lands server-side.
  final int? avalancheLevel;

  /// Short summary from Varsom. Null until the provider lands.
  final String? avalancheSummary;

  final int? score;
  final String rationale;

  const BackcountrySkiConditionsReport({
    required this.activityId,
    required this.validAt,
    required this.fetchedAt,
    required this.weather,
    required this.avalancheLevel,
    required this.avalancheSummary,
    required this.score,
    required this.rationale,
  });

  factory BackcountrySkiConditionsReport.fromJson(Map<String, dynamic> json) =>
      BackcountrySkiConditionsReport(
        activityId: json['activityId'] as String,
        validAt: DateTime.parse(json['validAt'] as String),
        fetchedAt: DateTime.parse(json['fetchedAt'] as String),
        weather: WeatherSlice.fromJson(json['weather'] as Map<String, dynamic>),
        avalancheLevel: (json['avalancheLevel'] as num?)?.toInt(),
        avalancheSummary: json['avalancheSummary'] as String?,
        score: (json['score'] as num?)?.toInt(),
        rationale: json['rationale'] as String,
      );
}

File: apps/api/src/Activities/Turbo.Activities.BackcountrySki.Contracts/value/BackcountrySkiConditionsReport.cs
using System.Text.Json.Serialization;
using Turboapi.Activities.value;

namespace Turboapi.Activities.BackcountrySki.value;

/// <summary>
/// Typed conditions report for a backcountry ski route. Carries the
/// same <see cref="WeatherSlice"/> the other kinds use, plus
/// kind-specific avalanche fields (currently always null pending a
/// real <c>VarsomAvalancheProvider</c> integration — the wire shape is
/// in place so the client can render the slot once data arrives).
/// </summary>
public sealed record BackcountrySkiConditionsReport
{
    [JsonPropertyName("activityId")] public Guid ActivityId { get; init; }
    [JsonPropertyName("validAt")] public DateTimeOffset ValidAt { get; init; }
    [JsonPropertyName("fetchedAt")] public DateTimeOffset FetchedAt { get; init; }

    [JsonPropertyName("weather")] public WeatherSlice Weather { get; init; }

    /// <summary>Varsom avalanche danger level 1–5 if available.
    /// Null until <c>VarsomAvalancheProvider</c> ships.</summary>
    [JsonPropertyName("avalancheLevel")] public int? AvalancheLevel { get; init; }

    /// <summary>Short summary string from Varsom. Null until provider
    /// ships.</summary>
    [JsonPropertyName("avalancheSummary")] public string? AvalancheSummary { get; init; }

    [JsonPropertyName("score")] public int? Score { get; init; }
    [JsonPropertyName("rationale")] public string Rationale { get; init; }

    [JsonConstructor]
    public BackcountrySkiConditionsReport(
        Guid activityId, DateTimeOffset validAt, DateTimeOffset fetchedAt,
        WeatherSlice weather,
        int? avalancheLevel, string? avalancheSummary,
        int? score, string rationale)
    {
        ActivityId = activityId;
        ValidAt = validAt;
        FetchedAt = fetchedAt;
        Weather = weather;
        AvalancheLevel = avalancheLevel;
        AvalancheSummary = avalancheSummary;
        Score = score;
        Rationale = rationale ?? throw new ArgumentNullException(nameof(rationale));
    }
}
